File: DevExpress VCL/ExpressQuantumGrid 6/Demos/Delphi/MasterDetailCardDemo/MasterDetailCardDemoMain.pas
unit MasterDetailCardDemoMain;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, cxGridCustomTableView, cxGridTableView, cxGridDBTableView, DB,
  DBTables, cxGridLevel, cxControls, cxGridCommon, cxGridCustomView, cxGrid,
  cxGridCardView, cxGridDBCardView, StdCtrls, ExtCtrls,
  cxDBData, ComCtrls, ToolWin, Menus, cxStyles, ImgList, DBCtrls,
  cxDropDownEdit, cxLookupEdit, cxDBLookupComboBox, cxMaskEdit, cxCalendar,
  cxDBEdit, cxHyperLinkEdit, cxContainer, cxEdit, cxTextEdit, cxMemo,
  ActnList, cxCheckBox, cxCustomData, cxGraphics, cxFilter, cxData,
  cxClasses, cxNavigator, cxDBLookupEdit, cxDBNavigator, cxDataStorage,
  cxBlobEdit, cxLookAndFeels;

type
  TMasterDetailCardDemoMainForm = class(TForm)
    Grid: TcxGrid;
    lvFilms: TcxGridLevel;
    pnlDetail: TPanel;
    tvFilms: TcxGridDBTableView;
    colFilmsCaption: TcxGridDBColumn;
    colFilmsYear: TcxGridDBColumn;
    colFilmsTagline: TcxGridDBColumn;
    colFilmsPlotOutline: TcxGridDBColumn;
    colFilmsRuntime: TcxGridDBColumn;
    colFilmsPhoto: TcxGridDBColumn;
    lblMaster: TLabel;
    lblDetail: TLabel;
    lblStyle: TLabel;
    Bevel1: TBevel;
    cvFilmsPersons: TcxGridDBCardView;
    cvFilmsPersonsPersonLineID: TcxGridDBCardViewRow;
    cvFilmsPersonsBIOGRAPHY: TcxGridDBCardViewRow;
    cvFilmsPersonsBIRTHNAME: TcxGridDBCardViewRow;
    cvFilmsPersonsDATEOFBIRTH: TcxGridDBCardViewRow;
    cvFilmsPersonsFIRSTNAME: TcxGridDBCardViewRow;
    cvFilmsPersonsLOCATIONOFBIRTH: TcxGridDBCardViewRow;
    cvFilmsPersonsNICKNAME: TcxGridDBCardViewRow;
    cvFilmsPersonsSECONDNAME: TcxGridDBCardViewRow;
    cvFilmsPersonsHOMEPAGE: TcxGridDBCardViewRow;
    mmMain: TMainMenu;
    miFile: TMenuItem;
    miExit: TMenuItem;
    miOptions: TMenuItem;
    miGrid4: TMenuItem;
    miAbout: TMenuItem;
    cvFilmsPersonsName: TcxGridDBCardViewRow;
    lbDescription: TLabel;
    StatusBar: TStatusBar;
    Panel1: TPanel;
    Label1: TLabel;
    cbOccupation: TcxDBLookupComboBox;
    Label2: TLabel;
    edFirstName: TcxDBTextEdit;
    Label8: TLabel;
    edSecondName: TcxDBTextEdit;
    Label9: TLabel;
    edNickName: TcxDBTextEdit;
    Label7: TLabel;
    edHomePage: TcxDBHyperLinkEdit;
    DBNavigator1: TcxDBNavigator;
    edBirthName: TcxDBTextEdit;
    Label6: TLabel;
    edLocationOfBirth: TcxDBTextEdit;
    Label4: TLabel;
    deDateOfBirth: TcxDBDateEdit;
    Label3: TLabel;
    meBiography: TcxDBMemo;
    Label5: TLabel;
    lvFilmsPersons: TcxGridLevel;
    chbMale: TcxDBCheckBox;
    cvFilmsPersonsGender: TcxGridDBCardViewRow;
    cxEditStyleController1: TcxEditStyleController;
    LookAndFeelController: TcxLookAndFeelController;
    procedure miAboutClick(Sender: TObject);
    procedure miExitClick(Sender: TObject);
    procedure miGrid4Click(Sender: TObject);
  private
    procedure SetStandardMasterDetailStyle;
    procedure SetGrid4MasterDetailStyle;
  public
    { Public declarations }
  end;

var
  MasterDetailCardDemoMainForm: TMasterDetailCardDemoMainForm;

implementation

{$R *.dfm}

uses
  cxGridDBDataDefinitions, MasterDetailCardDemoData, AboutDemoForm;

procedure TMasterDetailCardDemoMainForm.miExitClick(Sender: TObject);
begin
  Close;
end;

procedure TMasterDetailCardDemoMainForm.miAboutClick(Sender: TObject);
begin
  ShowAboutDemoForm;
end;

procedure TMasterDetailCardDemoMainForm.miGrid4Click(Sender: TObject);
begin
  with Sender as TMenuItem do
    Checked := not Checked;
  if (Sender as TMenuItem).Checked then
    SetGrid4MasterDetailStyle
  else
    SetStandardMasterDetailStyle;
end;

procedure TMasterDetailCardDemoMainForm.SetGrid4MasterDetailStyle;
var
  AData: TcxGridDBDataController;
begin
  pnlDetail.Visible := False;
  AData := cvFilmsPersons.DataController;
  AData.DataModeController.SmartRefresh := True;
  lvFilmsPersons.Visible := True;
  MasterDetailCardDemoDataDM.qryFilmsPersons.DataSource := nil;
  with MasterDetailCardDemoDataDM.qryFilmsPersons do
  begin
    Close;
    Params.Clear;
    SQL.Delete(SQL.Count-1);
    Open;
  end;
  lblMaster.Visible := False;
  lblStyle.Caption := 'ExpressQuantumGrid master-detail style';
end;

procedure TMasterDetailCardDemoMainForm.SetStandardMasterDetailStyle;
var
  AData: TcxGridDBDataController;
begin
  lvFilmsPersons.Visible := False;
  AData := cvFilmsPersons.DataController;
  AData.DataModeController.SmartRefresh := False;
  MasterDetailCardDemoDataDM.qryFilmsPersons.DataSource :=
    MasterDetailCardDemoDataDM.dsFilms;
  with MasterDetailCardDemoDataDM.qryFilmsPersons do
  begin
    Params.CreateParam(ftAutoInc, 'ID', ptUnknown);
    SQL.Add('and FilmID = :ID');
    Open;
  end;
  pnlDetail.Visible := True;
  lblMaster.Visible := True;
  lblStyle.Caption := 'Standard master-detail style';
end;

end.

File: DevExpress VCL/ExpressQuantumGrid 7/Demos/CBuilder/MasterDetailCardDemo/MasterDetailCardDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "shellapi.hpp"
#pragma hdrstop

#include "MasterDetailCardDemoMain.h"
#include "MasterDetailCardDemoData.h"
#include "AboutDemoForm.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxCalendar"
#pragma link "cxCheckBox"
#pragma link "cxContainer"
#pragma link "cxDBEdit"
#pragma link "cxDBLookupComboBox"
#pragma link "cxDropDownEdit"
#pragma link "cxGridCardView"
#pragma link "cxGridDBCardView"
#pragma link "cxHyperLinkEdit"
#pragma link "cxLookupEdit"
#pragma link "cxMaskEdit"
#pragma link "cxMemo"
#pragma link "cxTextEdit"
#pragma link "cxDBLookupEdit"
#pragma link "cxNavigator"
#pragma link "cxDBNavigator"
#pragma link "cxLookAndFeels"
#pragma resource "*.dfm"
TMasterDetailCardDemoMainForm *MasterDetailCardDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TMasterDetailCardDemoMainForm::TMasterDetailCardDemoMainForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TMasterDetailCardDemoMainForm::miAboutClick(TObject *Sender)
{
  ShowAboutDemoForm();
}
//---------------------------------------------------------------------------

void __fastcall TMasterDetailCardDemoMainForm::miExitClick(TObject *Sender)
{
  Close();
}
//---------------------------------------------------------------------------

void __fastcall TMasterDetailCardDemoMainForm::miGrid4Click(
      TObject *Sender)
{
  if (miGrid4->Checked)
    SetStandardMasterDetailStyle();
  else
    SetGrid4MasterDetailStyle();
  ((TMenuItem*)Sender)->Checked = !((TMenuItem*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void TMasterDetailCardDemoMainForm::SetStandardMasterDetailStyle(void)
{
  TcxGridDBDataController *AData;
  lvFilmsPersons->Visible = False;
  AData = cvFilmsPersons->DataController;
  AData->DataModeController->SmartRefresh = False;
  MasterDetailCardDemoDataDM->qryFilmsPersons->DataSource =
    MasterDetailCardDemoDataDM->dsFilms;
  MasterDetailCardDemoDataDM->qryFilmsPersons->Params->CreateParam(ftAutoInc, "ID", ptUnknown);
  MasterDetailCardDemoDataDM->qryFilmsPersons->SQL->Add("and FilmID = :ID");
  MasterDetailCardDemoDataDM->qryFilmsPersons->Open();
  pnlDetail->Visible = True;
  lblMaster->Visible = True;
  lblStyle->Caption = "Standard master-detail style";
}
//---------------------------------------------------------------------------

void TMasterDetailCardDemoMainForm::SetGrid4MasterDetailStyle(void)
{
  TcxGridDBDataController *AData;
  lvFilmsPersons->Visible = True;
  AData = cvFilmsPersons->DataController;
  AData->DataModeController->SmartRefresh = True;
  pnlDetail->Visible = False;
  MasterDetailCardDemoDataDM->qryFilmsPersons->DataSource = NULL;
  MasterDetailCardDemoDataDM->qryFilmsPersons->Close();
  MasterDetailCardDemoDataDM->qryFilmsPersons->Params->Clear();
  MasterDetailCardDemoDataDM->qryFilmsPersons->SQL->Delete(
    MasterDetailCardDemoDataDM->qryFilmsPersons->SQL->Count-1);
  MasterDetailCardDemoDataDM->qryFilmsPersons->Open();
  lblMaster->Visible = False;
  lblStyle->Caption = "ExpressQuantumGrid master-detail style";
}



